### ISSUE TITLE
How does FDBRepo.insert(entry.(), prefix: tenant) scale?

### ISSUE THREAD

**Issue by fire** [foundationdb-beam/ecto_foundationdb#4]
https://github.com/foundationdb-beam/ecto_foundationdb/blob/main/bench/scripts/insert_bench.exs

How does FDBRepo.insert(entry.(), prefix: tenant) scale?

For example:

```
##### With input Changeset #####
Name                         ips        average  deviation         median         99th %
FDB Repo.insert!/1        229.83        4.35 ms    ±18.73%        4.09 ms        6.37 ms

##### With input Struct #####
Name                         ips        average  deviation         median         99th %
FDB Repo.insert!/1        229.08        4.37 ms    ±19.35%        4.12 ms        6.45 ms
```

What can I do to double the ips?

**Comment by JesseStimpson** [comment]
Hi @fire, I suggest trying 

1. batching writes into a single transaction by using [Ecto.Schema.insert_all/3](https://hexdocs.pm/ecto/Ecto.Repo.html#c:insert_all/3)
2. performing multiple transactions in parallel. 

Both should get you performance gains, but the only way to make use of all database resources is by performing parallel transactions. Assuming you are inserting objects that all have unique primary keys, there will be no key contention. 

If you're going in-depth on scaling FDB itself, the [FoundationDB Performance doc](https://apple.github.io/foundationdb/performance.html) will probably be the best starting point for you.

There are a few things to consider specific to Ecto Adapters that result in a slight performance loss:

1. The `Ecto.Adapter.Schema.insert` and `insert_all` functions are expected to provide a guarantee that the object does not already exist in the database at the time of insert. EctoFDB implements this, but it does mean that the transaction waits for a get on each key (these gets are pipelined, but still noticeable).
2. There is an inventory of index metadata that is safely cached in memory during runtime. However, if the cache is not yet populated or if it gets invalidated due to a migration, there will be a small performance penalty on the next transaction. Subsequent transactions should make successful use of the cache. There is a small amount of overhead in maintaining the validity of the cache, but I expect it to be negligible.

**Comment by JesseStimpson** [comment]
FDB conflict ranges are checked in the context of in-flight transactions. They do not perform any `get` on keys to check for existence. My mental model for conflict ranges is similar to that of Read/Write mutex locks. Two parallel transactions are allowed to read from key A. But if one of those transactions has a write conflict specific for key A, then the transactions will be serialized by FDB.

This page does a much better job than me: [FDB Developer Guide | How FDB Detects Conflicts](https://apple.github.io/foundationdb/developer-guide.html#how-foundationdb-detects-conflicts)

This is different than a Postgres `ON CONFLICT` of course.

**Comment by JesseStimpson** [comment]
I've also just noticed this section for the first time, which provides concrete batching suggestions:

> Loading data
>
> Loading data is a common task in any database. Loading data in FoundationDB will be most efficiently accomplished if a > few guidelines are followed:
>
>    Load small sequential chunks of the data set from random positions in the data set (to allow the system to efficiently distribute different data to different servers).
>
>    Do about 10KB of data in total writes per transaction.
>
>    Use about 50 concurrent transactions per loading process to allow efficient pipelining. You can increase the number of concurrent transactions as long as transaction latencies remain under about 1 second.
>
>    Use multiple processes loading in parallel if a single one is CPU-bound.
>
> Using these techniques, our cluster of 24 nodes and 48 SSDs loads about 3 billion (100 byte) key-value pairs per hour.

(FDB Developer Guide | Loading Data)[https://apple.github.io/foundationdb/developer-guide.html#loading-data]

**Comment by fire** [comment]
Thanks for the pointers.

**Comment by jallum** [comment]
> FDB conflict ranges are checked in the context of in-flight transactions. They do not perform any `get` on keys to check for existence. My mental model for conflict ranges is similar to that of Read/Write mutex locks. Two parallel transactions are allowed to read from key A. But if one of those transactions has a write conflict specific for key A, then the transactions will be serialized by FDB.

Yep. Re-reading the foundationdb paper makes it clear that the shortcut I'm describing _would_ work, but only for transactions that share the same transaction window. Not useful here.

That said, I _did_ do some spelunking around in the code for your `insert_all`, and I _think_ it might be a missing a step. What's happening now is a _get_, followed by a _set_, if the value read is `:not_found`. I think you need to add an explicit call to `:erlfdb.add_read_conflict_key/2` before the _set_ to ensure "first write wins" semantics. I checked the source for erlfdb's nif to see if there was any magic in there, and no, it's just calling straight through to the c-library.

I could be wrong. 🤷

**Comment by JesseStimpson** [comment]
Thanks for linking the FDB forums discussion. The method described to perform the "is absent" check is an optimization to what's currently in EctoFDB. Please feel free to open a separate issue on that.

I don't believe the current implementation has a bug, however, because EctoFDB is not performing the `get` in question at snapshot isolation. Normal `get`s automatically get a read conflict key added to the transaction. Only if you request 'snapshot isolation' is the conflict key skipped.

What EctoFDB does today:
1. `get` key, not as snapshot (This adds the key to the read conflict range automatically)
2. If key is absent, `set` key
3. Commit

What the FDB Forum post suggests:
1. Read key at snapshot isolation
2. If key is absent, add key to read conflict range manually and set key to value
3. Commit

**Comment by jallum** [comment]
> I don't believe the current implementation has a bug, however, because EctoFDB is not performing the `get` in question at snapshot isolation. Normal `get`s automatically get a read conflict key added to the transaction. Only if you request 'snapshot isolation' is the conflict key skipped.

_oh_, that's interesting and subtle and was not at all clear (to me) from [the documentation](https://apple.github.io/foundationdb/api-c.html#snapshots) -- erlfdb's `get_key` and it's `_ss` variant are apparently calls to the same `fdb_transaction_get_key` under the hood. sorry for the tangent!

**Comment by JesseStimpson** [comment]
I'm going to close this issue for now, but please reach out if you have any other questions.
